From 0b73d6bbf91926cf9153c42ceaaf9f36f7c5181f Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Thu, 20 May 2021 20:34:27 +0300
Subject: [PATCH] [SW-36216] greco: security: hbw rr preparation

Implementation of base apis for hbw range register configuration.

There are 10(0..9 rr idx) RRs for HBW
NOTE: All the RRs for HBW are of 4KB granularity

1) 6(0..5 rr idx) Regions with 39 bit addressing (0x7F_Fx_xxxx)
API: preboot_secure_hbw_rr_short(rr idx, min addr, max addr)

2) 2(6..7 rr idx) Regions with 50 bit addressing.
API: preboot_secure_hbw_rr_full(rr idx, min addr, max addr)

3) 2(8..9 rr idx) Regions with 50 bit addressing, to block only write.
API: preboot_secure_hbw_rr_full(rr idx, min addr, max addr)

Change-Id: I7df51d2f6940745ba069c76a93793756f7a90120
Signed-off-by: Karthikeyan K <karthikk@habana.ai>
---

diff --git a/arch/greco/security.c b/arch/greco/security.c
index 59807ce..ca5a5e8 100644
--- a/arch/greco/security.c
+++ b/arch/greco/security.c
@@ -16,6 +16,14 @@
 #include <pcie_elbi_rr_mstr_if_rr_shrd_lbw_regs.h>
 #include <dcon0_lbw_rtr_if_mstr_if_rr_prvt_lbw_regs.h>
 #include <dcon0_lbw_rtr_if_mstr_if_rr_shrd_lbw_regs.h>
+#include <dcore0_tpcif_rtr0_mstr_if_rr_prvt_hbw_regs.h>
+#include <dcore1_tpcif_rtr0_mstr_if_rr_prvt_hbw_regs.h>
+#include <dcore0_mmeif_rtr0_mstr_if_rr_prvt_hbw_regs.h>
+#include <dcore1_mmeif_rtr0_mstr_if_rr_prvt_hbw_regs.h>
+#include <dcon0_hbw_rtr_if0_mstr_if_rr_prvt_hbw_regs.h>
+#include <dcon0_hbw_rtr_if0_mstr_if_rr_shrd_hbw_regs.h>
+#include <dcon0_hbw_rtr_if1_mstr_if_rr_prvt_hbw_regs.h>
+#include <dcon0_hbw_rtr_if1_mstr_if_rr_shrd_hbw_regs.h>
 
 #include "uart.h"
 
@@ -60,6 +68,10 @@
 	LBW_MAX_RR_IDX = 18,
 };
 
+enum greco_sec_if_hbw_rrs {
+	HBW_MAX_RR_IDX = 10,
+};
+
 #define EN_ARC_SEC_PRIV		0x3
 
 void preboot_secure_arc(void)
@@ -84,6 +96,10 @@
 #define COMMON_PRVT_DIST	0x10000
 #define COMMON_SHRD_NUM		2
 #define COMMON_SHRD_DIST	(COMMON_PRVT_DIST * 2)
+#define COMMON_HBW_PRVT_NUM	COMMON_PRVT_NUM
+#define COMMON_HBW_PRVT_DIST	COMMON_PRVT_DIST
+#define COMMON_HBW_SHRD_NUM	COMMON_SHRD_NUM
+#define COMMON_HBW_SHRD_DIST	COMMON_SHRD_DIST
 
 static struct preboot_sec_rrs_cfg preboot_sec_lbw_rrs_tbl[] = {
 	[TPC_MME_IF] = {
@@ -125,6 +141,69 @@
 		.max_full_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_0,
 		.hit_wr_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AW,
 		.hit_rd_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AR,
+	},
+};
+
+static struct preboot_sec_rrs_cfg preboot_sec_hbw_rrs_tbl[] = {
+	[TPC_MME_IF] = {
+		.instances_num = TPC_MME_IF_NUM,
+		.instances_dist = TPC_MME_IF_DIST,
+		.min_short_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MIN_SHORT_0,
+		.max_short_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MAX_SHORT_0,
+		.min_full_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MIN_LO_0,
+		.min_full_breg_hi = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MIN_HI_0,
+		.max_full_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MAX_LO_0,
+		.max_full_breg_hi = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MAX_HI_0,
+		.hit_wr_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_HBW_PRIV_HIT_AW,
+		.hit_rd_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_HBW_PRIV_HIT_AR,
+	},
+	[COMMON_IF_PRVT] = {
+		.instances_num = COMMON_HBW_PRVT_NUM,
+		.instances_dist = COMMON_HBW_PRVT_DIST,
+		.min_short_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MIN_SHORT_0,
+		.max_short_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MAX_SHORT_0,
+		.min_full_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MIN_LO_0,
+		.min_full_breg_hi = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MIN_HI_0,
+		.max_full_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MAX_LO_0,
+		.max_full_breg_hi = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MAX_HI_0,
+		.hit_wr_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_PRVT_HBW_PRIV_HIT_AW,
+		.hit_rd_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_PRVT_HBW_PRIV_HIT_AR,
+	},
+	[COMMON_IF_SHRD] = {
+		.instances_num = COMMON_HBW_PRVT_NUM,
+		.instances_dist = COMMON_HBW_PRVT_DIST,
+		.min_short_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_SHORT_0,
+		.max_short_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_SHORT_0,
+		.min_full_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0,
+		.min_full_breg_hi = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0,
+		.max_full_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0,
+		.max_full_breg_hi = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0,
+		.hit_wr_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_HIT_AW,
+		.hit_rd_breg = mmDCON0_HBW_RTR_IF0_MSTR_IF_RR_SHRD_HBW_PRIV_HIT_AR,
+	},
+	[ELBI_OR_COMMON_IF2] = {
+		.instances_num = COMMON_HBW_PRVT_NUM,
+		.instances_dist = COMMON_HBW_PRVT_DIST,
+		.min_short_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MIN_SHORT_0,
+		.max_short_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MAX_SHORT_0,
+		.min_full_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MIN_LO_0,
+		.min_full_breg_hi = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MIN_HI_0,
+		.max_full_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MAX_LO_0,
+		.max_full_breg_hi = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_PRVT_HBW_PRIV_RANGE_MAX_HI_0,
+		.hit_wr_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_PRVT_HBW_PRIV_HIT_AW,
+		.hit_rd_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_PRVT_HBW_PRIV_HIT_AR,
+	},
+	[COMMON_IF_SHRD2] = {
+		.instances_num = COMMON_HBW_PRVT_NUM,
+		.instances_dist = COMMON_HBW_PRVT_DIST,
+		.min_short_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_SHORT_0,
+		.max_short_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_SHORT_0,
+		.min_full_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_LO_0,
+		.min_full_breg_hi = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MIN_HI_0,
+		.max_full_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_LO_0,
+		.max_full_breg_hi = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_RANGE_MAX_HI_0,
+		.hit_wr_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_HIT_AW,
+		.hit_rd_breg = mmDCON0_HBW_RTR_IF1_MSTR_IF_RR_SHRD_HBW_PRIV_HIT_AR,
 	},
 };
 
@@ -210,6 +289,77 @@
 	preboot_secure_lbw_rr_wrap(rr_idx, min, max, true);
 }
 
+#define MAX_HBW_SHORT_RR	6
+#define MAX_HBW_READ_RRS	7
+#define BITS_IN_REG		32
+
+static void preboot_secure_hbw_rr_full_if(int rr_idx, u64 min64, u64 max64,
+		enum preboot_sec_engine_type type)
+{
+	u32 min_l = (u32)(min64);
+	u32 min_h = (u32)(min64 >> BITS_IN_REG);
+	u32 max_l = (u32)(max64);
+	u32 max_h = (u32)(max64 >> BITS_IN_REG);
+	u32 rr_base, rr_dest, rr_calc;
+	int i;
+	struct preboot_sec_rrs_cfg *cfg = &preboot_sec_hbw_rrs_tbl[type];
+
+	rr_calc = rr_idx - MAX_HBW_SHORT_RR;
+	for (i = 0; i < cfg->instances_num; i++) {
+		rr_base = rr_calc * sizeof(u32) + cfg->instances_dist * i;
+		rr_dest = rr_base + cfg->min_full_breg;
+		regmap_writel(min_l, rr_dest);
+		rr_dest = rr_base + cfg->min_full_breg_hi;
+		regmap_writel(min_h, rr_dest);
+		rr_dest = rr_base + cfg->max_full_breg;
+		regmap_writel(max_l, rr_dest);
+		rr_dest = rr_base + cfg->max_full_breg_hi;
+		regmap_writel(max_h, rr_dest);
+
+		/*
+		 * RRs 8..9 only blocks write operations
+		 */
+		if (rr_idx > MAX_HBW_READ_RRS)
+			preboot_secure_update_hit_reg(rr_idx, cfg, i, true);
+		else
+			preboot_secure_update_hit_reg(rr_idx, cfg, i, false);
+	}
+}
+
+static void preboot_secure_hbw_rr_short_if(int rr_idx, u32 min, u32 max,
+		enum preboot_sec_engine_type type)
+{
+	int i;
+	struct preboot_sec_rrs_cfg *cfg = &preboot_sec_hbw_rrs_tbl[type];
+
+	for (i = 0; i < cfg->instances_num; i++) {
+		preboot_secure_update_rr_reg(rr_idx, cfg->instances_dist * i, cfg->min_short_breg,
+				cfg->max_short_breg, min, max);
+
+		preboot_secure_update_hit_reg(rr_idx, cfg, i, false);
+	}
+}
+
+__unused static void preboot_secure_hbw_rr_full(int rr, u64 min64, u64 max64)
+{
+	u64 min = min64 >> SHIFT_4K_POS;
+	u64 max = max64 >> SHIFT_4K_POS;
+	int i;
+
+	for (i = TPC_MME_IF; i < COMMON_IF_SHRD2; i++)
+		preboot_secure_hbw_rr_full_if(rr, min, max, (enum preboot_sec_engine_type)i);
+}
+
+__unused static void preboot_secure_hbw_rr_short(int rr, u64 min64, u64 max64)
+{
+	u32 min = (min64 & ~SDRAM_BASE) >> SHIFT_4K_POS;
+	u32 max = (max64 & ~SDRAM_BASE) >> SHIFT_4K_POS;
+	int i;
+
+	for (i = TPC_MME_IF; i < COMMON_IF_SHRD2; i++)
+		preboot_secure_hbw_rr_short_if(rr, min, max, (enum preboot_sec_engine_type)i);
+}
+
 static void preboot_secure_lbw(void)
 {
 	/*
@@ -228,9 +378,29 @@
 	 */
 }
 
+static void preboot_secure_hbw(void)
+{
+	/*
+	 * There are 10(0..9 rr idx) RRs for HBW
+	 * NOTE: All the RRs for HBW are of 4KB granularity
+	 *
+	 * 1) 6(0..5 rr idx) Regions with 39 bit addressing (0x7F_Fx_xxxx)
+	 * API: preboot_secure_hbw_rr_short(rr idx, min addr, max addr)
+	 *
+	 * 2) 2(6..7 rr idx) Regions with 50 bit addressing.
+	 * API: preboot_secure_hbw_rr_full(rr idx, min addr, max addr)
+	 *
+	 * 3) 2(8..9 rr idx) Regions with 50 bit addressing, to block only write.
+	 * API: preboot_secure_hbw_rr_full(rr idx, min addr, max addr)
+	 *
+	 * rr idx type : greco_sec_if_hbw_rrs
+	 */
+}
+
 void preboot_security_init(void)
 {
 	/* TODO add check for security strap ON */
 	preboot_secure_arc();
 	preboot_secure_lbw();
+	preboot_secure_hbw();
 }
